cachi2/core/package_managers/yarn_classic/project.py

```diff
@@ -0,0 +1,294 @@
+                solution=(
+                    "Please double-check that you have specified the correct path "
+                    "to the package directory containing this file"
+                ),
+            )
+        except json.decoder.JSONDecodeError as e:
+            raise PackageRejected(
+                reason=f"Can't parse the {path.subpath_from_root} file. {e}",
+                solution=(
+                    "The package.json file must contain valid JSON. "
+                    "Refer to the parser error and fix the contents of the file."
+                ),
+            )
+
+        return cls(path, package_json_data)
+
+
+class YarnRc(_CommonConfigFile):
+    """A yarnrc file."""
+
+    _cache_folder: Optional[str]
+
+    @property
+    def cache_folder(self) -> str:
+        """Get the configured location for the yarn cache folder.
+
+        Fallback to the default path in case the configuration key is missing.
+        """
+        return self._data.get("--cache-folder", DEFAULT_CACHE_FOLDER)
+
+    @classmethod
+    def from_file(cls, path: RootedPath) -> "YarnRc":
+        """Parse the content of a yarnrc file.
+
+        .yarnrc file are seriously dumb, just plain text keys and values with a
+        single space in between. According to
+        https://classic.yarnpkg.com/en/docs/yarnrc, paths should be quoted
+
+        `yarn-offline-mirror "./packages-cache"`
+
+        CLI syntax is possible
+
+        `--install.check-files true`
+
+        but paths don't have to be quoted(?!)
+
+        `--cache-folder /tmp/yarn-cache/`
+        """
+
+        yarnrc_data = {}
+        try:
+            with path.path.open("r") as f:
+                for line in f:
+                    line = line.replace("\n", "").strip()
+                    if line:
+                        key_and_val = line.split(" ", 1)
+                        yarnrc_data[key_and_val[0]] = key_and_val[1]
+        except ValueError as e:
+            raise PackageRejected(
+                reason=f"Can't parse the {path.subpath_from_root} file. Parser error: {e}",
+                solution=(
+                    "The yarnrc file must contain valid data."
+                    "Refer to the parser error and fix the contents of the file."
+                ),
+            )
+
+        if not yarnrc_data:
+            # warn
+            pass
+
+        return cls(path, yarnrc_data)
+
+    @classmethod
+    def from_str(cls, string: str) -> dict[str, Any]:
+        """Parse the content of a string containing the contents of a valid yarnrc file."""
+        yarnrc_data = {}
+        try:
+            for line in string.splitlines():
+                line.strip()
+                if line:
+                    key_and_val = line.split(" ", 1)
+                    yarnrc_data[key_and_val[0]] = key_and_val[1]
+        except ValueError as e:
+            raise PackageRejected(
```

Comment:
I see what you're saying; this is the pattern I copied from Yarnberry. If we have an error parsing any config file *for the current package*, we can't continue, so `PackageRejected` is appropriate.

Isn't it? :laughing:

Reply:
From a purist's perspective we should be raising ParsingError here which must be converted to PackageRejected at caller's level. I am not a purist and won't block on this.

Reply:
I have not checked yet, but I have a strong suspicion that this would break the existing pattern. Refactoring half of PMs is definitely not in scope, so we likely can keep it as is for now.